File: dbgen/p02_fraud.py
import duckdb, numpy as np, sys, os
from datetime import datetime, timedelta
from concurrent.futures import ProcessPoolExecutor
from utils.synth_utils import (
    GenerationProgress,
    batched_insert,
    get_worker_count,
    print_generation_summary,
    run_parallel,
)


def generate_accounts_chunk(start, end, atypes, base):
    size = end - start
    rng = np.random.default_rng(start)
    atype_indices = rng.integers(0, len(atypes), size)
    country_indices = rng.integers(0, 5, size)
    countries = ["US", "GB", "DE", "FR", "CA"]
    limits = rng.uniform(500, 50000, size)
    days_back = rng.integers(30, 3651, size)
    frozen_probs = rng.random(size)
    
    rows = []
    for idx, i in enumerate(range(start, end)):
        rows.append((
            i,
            f"Holder {i}",
            atypes[atype_indices[idx]],
            countries[country_indices[idx]],
            round(float(limits[idx]), 2),
            (base - timedelta(days=int(days_back[idx]))).date(),
            bool(frozen_probs[idx] < 0.03),
        ))
    return rows

def generate_merchants_chunk(start, end, cats, risks):
    size = end - start
    rng = np.random.default_rng(start)
    cat_indices = rng.integers(0, len(cats), size)
    country_indices = rng.integers(0, 5, size)
    countries = ["US", "GB", "NG", "CN", "RU"]
    risk_indices = rng.integers(0, len(risks), size)
    avg_amounts = rng.uniform(5, 500, size)

    rows = []
    for idx, i in enumerate(range(start, end)):
        rows.append((
            i,
            f"Merchant {i}",
            cats[cat_indices[idx]],
            countries[country_indices[idx]],
            risks[risk_indices[idx]],
            round(float(avg_amounts[idx]), 2),
        ))
    return rows

def generate_transactions_chunk(start, end, NA, NM, base, chans, currs, rcodes):
    size = end - start
    rng = np.random.default_rng(start)
    account_ids = rng.integers(1, NA + 1, size)
    merchant_ids = rng.integers(1, NM + 1, size)
    amounts = rng.uniform(1, 5000, size)
    seconds_offset = rng.integers(0, 365 * 86400 + 1, size)
    chan_indices = rng.integers(0, len(chans), size)
    curr_indices = rng.integers(0, len(currs), size)
    declined_probs = rng.random(size)
    flagged_probs = rng.random(size)
    rcode_indices = rng.integers(0, len(rcodes), size)

    rows = []
    for idx, i in enumerate(range(start, end)):
        rows.append((
            i,
            int(account_ids[idx]),
            int(merchant_ids[idx]),
            round(float(amounts[idx]), 2),
            base + timedelta(seconds=int(seconds_offset[idx])),
            chans[chan_indices[idx]],
            currs[curr_indices[idx]],
            bool(declined_probs[idx] < 0.05),
            bool(flagged_probs[idx] < 0.04),
            rcodes[rcode_indices[idx]],
        ))
    return rows

def generate_alerts_chunk(start, end, flagged_ids, NT, atypes2, sevs, base, ress):
    size = end - start
    rng = np.random.default_rng(start)
    
    if flagged_ids:
        flagged_indices = rng.integers(0, len(flagged_ids), size)
    else:
        random_txn_ids = rng.integers(1, NT + 1, size)
        
    atype2_indices = rng.integers(0, len(atypes2), size)
    sev_indices = rng.integers(0, len(sevs), size)
    seconds_offset = rng.integers(0, 365 * 86400 + 1, size)
    resolved_probs = rng.random(size)
    res_probs = rng.random(size)
    res_indices = rng.integers(0, len(ress), size)

    rows = []
    for idx, i in enumerate(range(start, end)):
        txn_id = int(flagged_ids[flagged_indices[idx]]) if flagged_ids else int(random_txn_ids[idx])
        rows.append((
            i,
            txn_id,
            atypes2[atype2_indices[idx]],
            sevs[sev_indices[idx]],
            base + timedelta(seconds=int(seconds_offset[idx])),
            bool(resolved_probs[idx] > 0.4),
            ress[res_indices[idx]] if res_probs[idx] > 0.4 else None,
        ))
    return rows


def main():
    sf = float(sys.argv[1]) if len(sys.argv) > 1 else 1.0
    sf_adj = sf * 2200.0
    NA = max(10, int(1000 * sf_adj))
    NM = max(10, int(300 * sf_adj))
    NT = max(50, int(20000 * sf_adj))
    NAL = max(5, int(500 * sf_adj))


    os.makedirs("data", exist_ok=True)
    con = duckdb.connect("data/warehouse.duckdb")

    con.execute("""
    DROP TABLE IF EXISTS alerts; DROP TABLE IF EXISTS transactions;
    DROP TABLE IF EXISTS merchants; DROP TABLE IF EXISTS accounts;
    CREATE TABLE accounts(account_id INTEGER PRIMARY KEY, holder_name VARCHAR,
        account_type VARCHAR, country VARCHAR, credit_limit DECIMAL(12,2),
        opened_date DATE, is_frozen BOOLEAN);
    CREATE TABLE merchants(merchant_id INTEGER PRIMARY KEY, name VARCHAR,
        category VARCHAR, country VARCHAR, risk_tier VARCHAR,
        avg_txn_amount DECIMAL(10,2));
    CREATE TABLE transactions(txn_id INTEGER PRIMARY KEY, account_id INTEGER,
        merchant_id INTEGER, amount DECIMAL(12,2), txn_ts TIMESTAMP,
        channel VARCHAR, currency VARCHAR, is_declined BOOLEAN,
        is_flagged BOOLEAN, response_code VARCHAR);
    CREATE TABLE alerts(alert_id INTEGER PRIMARY KEY, txn_id INTEGER,
        alert_type VARCHAR, severity VARCHAR, created_ts TIMESTAMP,
        resolved BOOLEAN, resolution VARCHAR);
    """)

    base = datetime(2022, 1, 1)
    atypes = ["checking", "savings", "credit", "business"]
    cats = ["retail", "travel", "grocery", "online", "gaming", "crypto", "atm"]
    risks = ["low", "medium", "high", "critical"]
    chans = ["pos", "web", "mobile", "atm", "wire"]
    currs = ["USD", "EUR", "GBP", "JPY", "BTC"]
    rcodes = ["00", "01", "05", "14", "51", "57", "96"]
    atypes2 = ["velocity", "geo_anomaly", "amount_spike", "card_not_present", "identity"]
    sevs = ["info", "warning", "critical"]
    ress = ["confirmed_fraud", "false_positive", "under_review"]

    cpu_count = get_worker_count()
    progress = GenerationProgress("p02_fraud", 4)
    with ProcessPoolExecutor(max_workers=cpu_count) as executor:
        progress.advance("accounts")
        batched_insert(con, "accounts", ['account_id', 'holder_name', 'account_type', 'country', 'credit_limit', 'opened_date', 'is_frozen'],
                       run_parallel(executor, generate_accounts_chunk, NA, atypes, base))
        progress.advance("merchants")
        batched_insert(con, "merchants", ['merchant_id', 'name', 'category', 'country', 'risk_tier', 'avg_txn_amount'],
                       run_parallel(executor, generate_merchants_chunk, NM, cats, risks))
        
        progress.advance("transactions")
        txns = run_parallel(executor, generate_transactions_chunk, NT, NA, NM, base, chans, currs, rcodes)
        batched_insert(con, "transactions", ['txn_id', 'account_id', 'merchant_id', 'amount', 'txn_ts', 'channel', 'currency', 'is_declined', 'is_flagged', 'response_code'], txns)

        flagged_ids = [
            row[0]
            for row in con.execute(
                "SELECT txn_id FROM transactions WHERE is_flagged ORDER BY txn_id LIMIT ?",
                [NAL],
            ).fetchall()
        ]
        progress.advance("alerts")
        batched_insert(con, "alerts", ['alert_id', 'txn_id', 'alert_type', 'severity', 'created_ts', 'resolved', 'resolution'],
                       run_parallel(executor, generate_alerts_chunk, NAL, flagged_ids, NT, atypes2, sevs, base, ress))


    con.close()
    print_generation_summary(
        "p02_fraud",
        sf,
        {
            "accounts": NA,
            "merchants": NM,
            "transactions": NT,
            "alerts": NAL,
        },
    )

if __name__ == "__main__":
    main()

File: dbgen-rust/src/p02_fraud.rs
use chrono::{Duration, NaiveDate};
use duckdb::DuckdbConnectionManager;
use indicatif::{ProgressBar, ProgressStyle};
use r2d2::Pool;
use rand::prelude::*;
use rand::rngs::SmallRng;

pub fn run(sf: f64, pool: &mut Pool<DuckdbConnectionManager>) -> duckdb::Result<()> {
    let sf_adj = sf * 2200.0;
    let na = (1000.0 * sf_adj).max(10.0) as usize;
    let nm = (300.0 * sf_adj).max(10.0) as usize;
    let nt = (20000.0 * sf_adj).max(50.0) as usize;
    let nal = (500.0 * sf_adj).max(5.0) as usize;

    let con = &pool.get().expect("couldn't get connection");

    con.execute_batch(
        "DROP TABLE IF EXISTS alerts; DROP TABLE IF EXISTS transactions;
         DROP TABLE IF EXISTS merchants; DROP TABLE IF EXISTS accounts;
         CREATE TABLE accounts(account_id INTEGER PRIMARY KEY, holder_name VARCHAR,
             account_type VARCHAR, country VARCHAR, credit_limit DECIMAL(12,2),
             opened_date DATE, is_frozen BOOLEAN);
         CREATE TABLE merchants(merchant_id INTEGER PRIMARY KEY, name VARCHAR,
             category VARCHAR, country VARCHAR, risk_tier VARCHAR,
             avg_txn_amount DECIMAL(10,2));
         CREATE TABLE transactions(txn_id INTEGER PRIMARY KEY, account_id INTEGER,
             merchant_id INTEGER, amount DECIMAL(12,2), txn_ts TIMESTAMP,
             channel VARCHAR, currency VARCHAR, is_declined BOOLEAN,
             is_flagged BOOLEAN, response_code VARCHAR);
         CREATE TABLE alerts(alert_id INTEGER PRIMARY KEY, txn_id INTEGER,
             alert_type VARCHAR, severity VARCHAR, created_ts TIMESTAMP,
             resolved BOOLEAN, resolution VARCHAR);",
    )?;

    let base_ts = NaiveDate::from_ymd_opt(2022, 1, 1)
        .unwrap()
        .and_hms_opt(0, 0, 0)
        .unwrap();
    let atypes = ["checking", "savings", "credit", "business"];
    let cats = [
        "retail", "travel", "grocery", "online", "gaming", "crypto", "atm",
    ];
    let risks = ["low", "medium", "high", "critical"];
    let chans = ["pos", "web", "mobile", "atm", "wire"];
    let currs = ["USD", "EUR", "GBP", "JPY", "BTC"];
    let rcodes = ["00", "01", "05", "14", "51", "57", "96"];
    let atypes2 = [
        "velocity",
        "geo_anomaly",
        "amount_spike",
        "card_not_present",
        "identity",
    ];
    let sevs = ["info", "warning", "critical"];
    let ress = ["confirmed_fraud", "false_positive", "under_review"];
    let countries_acc = ["US", "GB", "DE", "FR", "CA"];
    let countries_merch = ["US", "GB", "NG", "CN", "RU"];

    let pb = ProgressBar::new(4);
    pb.set_style(
        ProgressStyle::default_bar()
            .template("[{elapsed_precise}] {bar:40.cyan/blue} {pos}/{len} {msg}")
            .unwrap(),
    );

    // 1. Accounts
    crate::generate_table_parallel(con, "accounts", na, &pb, "Generating accounts...", |i| {
        let mut rng = SmallRng::seed_from_u64(i as u64);
        let holder_name = format!("Holder {}", i);
        let atype = atypes[rng.gen_range(0..atypes.len())];
        let country = countries_acc[rng.gen_range(0..countries_acc.len())];
        let limit = ((rng.gen_range(500.0..50000.0) * 100.0) as f64).round() / 100.0;
        let opened_date = (base_ts - Duration::days(rng.gen_range(30..3651))).date();
        let is_frozen = rng.gen_bool(0.03);
        (
            i as i32,
            holder_name,
            atype,
            country,
            limit,
            opened_date,
            is_frozen,
        )
    })?;

    // 2. Merchants
    crate::generate_table_parallel(con, "merchants", nm, &pb, "Generating merchants...", |i| {
        let mut rng = SmallRng::seed_from_u64(i as u64);
        let name = format!("Merchant {}", i);
        let cat = cats[rng.gen_range(0..cats.len())];
        let country = countries_merch[rng.gen_range(0..countries_merch.len())];
        let risk = risks[rng.gen_range(0..risks.len())];
        let avg_amount = ((rng.gen_range(5.0..500.0) * 100.0) as f64).round() / 100.0;
        (i as i32, name, cat, country, risk, avg_amount)
    })?;

    // 3. Transactions
    crate::generate_table_parallel(
        con,
        "transactions",
        nt,
        &pb,
        "Generating transactions...",
        |i| {
            let mut rng = SmallRng::seed_from_u64(i as u64);
            let acc_id = rng.gen_range(1..=na) as i32;
            let merch_id = rng.gen_range(1..=nm) as i32;
            let amount = ((rng.gen_range(1.0..5000.0) * 100.0) as f64).round() / 100.0;
            let ts = base_ts + Duration::seconds(rng.gen_range(0..365 * 86400));
            let channel = chans[rng.gen_range(0..chans.len())];
            let curr = currs[rng.gen_range(0..currs.len())];
            let declined = rng.gen_bool(0.05);
            let flagged = rng.gen_bool(0.04);
            let rcode = rcodes[rng.gen_range(0..rcodes.len())];
            (
                i as i32, acc_id, merch_id, amount, ts, channel, curr, declined, flagged, rcode,
            )
        },
    )?;

    // Get flagged IDs for alerts
    let mut stmt =
        con.prepare("SELECT txn_id FROM transactions WHERE is_flagged ORDER BY txn_id LIMIT ?")?;
    let flagged_ids: Vec<i32> = stmt
        .query_map([nal as i32], |row| row.get(0))?
        .collect::<Result<Vec<i32>, _>>()?;

    // 4. Alerts
    crate::generate_table_parallel(con, "alerts", nal, &pb, "Generating alerts...", |i| {
        let mut rng = SmallRng::seed_from_u64(i as u64);
        let txn_id = if !flagged_ids.is_empty() {
            flagged_ids[rng.gen_range(0..flagged_ids.len())]
        } else {
            rng.gen_range(1..=nt) as i32
        };
        let atype2 = atypes2[rng.gen_range(0..atypes2.len())];
        let sev = sevs[rng.gen_range(0..sevs.len())];
        let ts = base_ts + Duration::seconds(rng.gen_range(0..365 * 86400));
        let resolved = rng.gen_bool(0.6);
        let res = if rng.gen_bool(0.6) {
            Some(ress[rng.gen_range(0..ress.len())])
        } else {
            None
        };
        (i as i32, txn_id, atype2, sev, ts, resolved, res)
    })?;

    pb.finish_with_message("p02_fraud complete");

    Ok(())
}
